### PR TITLE
javaslang.collection.Vector #464

### PR DESCRIPTION
**Pull request by ruslansennov**

### CODE REVIEW

File: src/main/java/javaslang/collection/Vector.java [hunk 1]
```diff
@@ -0,0 +1,1481 @@
...
+                vector = new Impl<>(temp);
+            }
+
+            private Object readResolve() {
+                return vector;
+            }
+        }
+    }
+
+    class Nil<T> implements Vector<T>, Serializable {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: And _boom_, Ruslan comes with Vector around the corner :-)) I love this HAMT!!

I've learned that Nil is somewhat 'reserved' for the empty (lazy) linked list. Lets name it `Empty`.

File: src/main/java/javaslang/collection/Vector.java [hunk 2]
```diff
@@ -0,0 +1,1481 @@
...
+            return Traversable.hash(this);
+        }
+
+        @Override
+        public String toString() {
+            return "Vector()";
+        }
+
+        @Override
+        public HashArrayMappedTrie<Integer, T> toHashArrayMappedTrie() {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: This is an implementation detail. A vector may be backed by another data structure. Let's make it not public API and remove this method.

File: src/main/java/javaslang/collection/Vector.java [hunk 3]
```diff
@@ -0,0 +1,1481 @@
...
+                return vector;
+            }
+        }
+    }
+
+    class Nil<T> implements Vector<T>, Serializable {
+        private static final long serialVersionUID = 1L;
+        private static final Vector<?> INSTANCE = new Nil<>();
+
+        private final HashArrayMappedTrie<Integer, T> trie = HashArrayMappedTrie.empty();
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: if we really need the trie instance in another place, let's make it package private but for now leave it private if possible.

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: Or better: remove it if possible. An empty Vector should be identified by `isEmpty()`.

File: src/main/java/javaslang/collection/Vector.java (hunk 1 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: Let's make it public.

File: src/main/java/javaslang/collection/Vector.java (hunk 2 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/Vector.java]: I need <code>Vector.toHashArrayMappedTrie()</code> method to be able to implement 99% of code in interface body. This hack is not so beautiful but very effective

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: I see. I would prefer to implement all methods for the empty case in the `Empty` instance. But that's a huge amount of extra work. For now let's keep it this way. It's ok.

Comment (ruslansennov) [inline src/main/java/javaslang/collection/Vector.java]: No, I'll do it :)

File: src/main/java/javaslang/collection/Vector.java (hunk 1 above)
Comment (ruslansennov) [inline src/main/java/javaslang/collection/Vector.java]: It's public :)

File: src/main/java/javaslang/collection/Vector.java (hunk 2 above)
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: That's the very thing! We need the abstract interface though. Bonus: We can adjust the return types to `Impl` and `Empty` where possible.

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: In exchange I will not commit & push until you are finished :-) So if there's s.th. to merge I'll do it!

File: src/main/java/javaslang/collection/Vector.java [hunk 4]
```diff
@@ -0,0 +1,1481 @@
...
+import java.io.*;
+import java.util.*;
+import java.util.function.*;
+import java.util.stream.Collector;
+
+/**
+ * TODO javadoc
+ * @param <T>
+ */
+public interface Vector<T> extends IndexedSeq<T> {
```
Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: Beside the trie method all other things look awesome to me!

Comment (ruslansennov) [inline src/main/java/javaslang/collection/Vector.java]: What do you think about <code>Vector.ofTrie()</code>?

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: Please remove all public API related to Trie. We do not really need it. The new `Value` I committed yesterday has all conversion methods between 'regular' collection one could need. I see HAMT, RedBlackTree, ... as backing collections, not as core collections. They do not extends Traversable et al. If in doubt, we have implemented the Iterable interface, which is sufficient to create any Traversable collection via `ofAll`.

Comment (ruslansennov) [inline src/main/java/javaslang/collection/Vector.java]: Well

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: Hold fire! HashMap, HashSet and Queue are currently classes. I thought about making them interfaces. But we may also stay with them as classes and make Vector a class instead. Then you could keep your implementation and remove the inner classes.

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: It needs to be `final`.

Comment (danieldietrich) [inline src/main/java/javaslang/collection/Vector.java]: And the trie stuff can be private...

**Comment by ruslansennov** [comment]
done

**Comment by danieldietrich** [comment]
Thx!

**Pull request merged**